File: apps/api/src/my_bot_api/auth/otp.py
from dataclasses import dataclass

from redis.asyncio import Redis

from my_bot_api.auth.errors import AuthError, invalid_code_error
from my_bot_api.auth.security import (
    generate_opaque_token,
    generate_otp_code,
    hash_otp,
    keyed_identifier,
)
from my_bot_api.config import Settings
from my_bot_api.services.email import EmailDeliveryError, OtpEmailSender

_FIXED_WINDOW_SCRIPT = """
local current = redis.call('INCR', KEYS[1])
if current == 1 then
  redis.call('EXPIRE', KEYS[1], ARGV[1])
end
local ttl = redis.call('TTL', KEYS[1])
return {current, ttl}
"""

_ROLLBACK_ISSUE_SCRIPT = """
if redis.call('GET', KEYS[1]) == ARGV[1] then
  redis.call('DEL', KEYS[1])
end
if redis.call('GET', KEYS[2]) == ARGV[1] then
  redis.call('DEL', KEYS[2])
end
redis.call('DEL', KEYS[3])
return 1
"""

_INSTALL_ISSUE_SCRIPT = """
local current = redis.call('GET', KEYS[1])
if ARGV[1] == '__none__' then
  if current then return {'stale'} end
elseif current ~= ARGV[1] then
  return {'stale'}
end
if current then redis.call('DEL', KEYS[3]) end
redis.call(
  'HSET', KEYS[2], 'code_hash', ARGV[3], 'email', ARGV[4],
  'attempts', '0', 'status', 'active'
)
redis.call('EXPIRE', KEYS[2], ARGV[5])
redis.call('SET', KEYS[1], ARGV[2], 'EX', ARGV[5])
return {'installed'}
"""

_RESERVE_SCRIPT = """
if redis.call('EXISTS', KEYS[1]) == 0 then return {'missing'} end
if redis.call('GET', KEYS[2]) ~= ARGV[1] then return {'missing'} end
if redis.call('HGET', KEYS[1], 'status') == 'reserved' then return {'already_reserved'} end
local expected = redis.call('HGET', KEYS[1], 'code_hash')
if expected == ARGV[2] then
  local email = redis.call('HGET', KEYS[1], 'email')
  redis.call('HSET', KEYS[1], 'status', 'reserved', 'reservation', ARGV[4])
  return {'verified', email}
end
local attempts = redis.call('HINCRBY', KEYS[1], 'attempts', 1)
if attempts >= tonumber(ARGV[3]) then
  redis.call('DEL', KEYS[1])
  redis.call('DEL', KEYS[2])
  return {'locked'}
end
return {'invalid', tostring(attempts)}
"""

_FINALIZE_SCRIPT = """
if redis.call('HGET', KEYS[1], 'reservation') ~= ARGV[1] then return 0 end
redis.call('DEL', KEYS[1])
if redis.call('GET', KEYS[2]) == ARGV[2] then redis.call('DEL', KEYS[2]) end
return 1
"""

_RELEASE_SCRIPT = """
if redis.call('HGET', KEYS[1], 'reservation') ~= ARGV[1] then return 0 end
if redis.call('GET', KEYS[2]) == ARGV[2] then
  redis.call('HSET', KEYS[1], 'status', 'active')
  redis.call('HDEL', KEYS[1], 'reservation')
else
  redis.call('DEL', KEYS[1])
end
return 1
"""


@dataclass(frozen=True, slots=True)
class IssuedOtp:
    challenge_id: str
    expires_in_seconds: int
    resend_after_seconds: int


@dataclass(frozen=True, slots=True)
class OtpReservation:
    challenge_id: str
    email: str
    reservation_id: str


class OtpService:
    def __init__(self, redis: Redis, email_sender: OtpEmailSender, settings: Settings) -> None:
        self._redis = redis
        self._email_sender = email_sender
        self._settings = settings

    def _email_key(self, email: str) -> str:
        return keyed_identifier(
            email,
            pepper=self._settings.rate_limit_pepper.get_secret_value(),
        )

    def _ip_key(self, ip_address: str) -> str:
        return keyed_identifier(
            ip_address,
            pepper=self._settings.rate_limit_pepper.get_secret_value(),
        )

    async def _take_limit(self, key: str, maximum: int) -> None:
        current, ttl = await self._redis.eval(
            _FIXED_WINDOW_SCRIPT,
            1,
            key,
            self._settings.otp_rate_limit_window_seconds,
        )
        if int(current) <= maximum:
            return
        raise AuthError(
            code="rate_limited",
            message="Too many attempts. Wait before trying again.",
            status_code=429,
            retry_after_seconds=max(1, int(ttl)),
        )

    async def issue(self, *, email: str, ip_address: str) -> IssuedOtp:
        email_key = self._email_key(email)
        challenge_id = generate_opaque_token()
        cooldown_key = f"auth:otp:cooldown:{email_key}"
        cooldown_acquired = await self._redis.set(
            cooldown_key,
            challenge_id,
            ex=self._settings.otp_resend_cooldown_seconds,
            nx=True,
        )
        if not cooldown_acquired:
            ttl = await self._redis.ttl(cooldown_key)
            raise AuthError(
                code="rate_limited",
                message="Wait before requesting another code.",
                status_code=429,
                retry_after_seconds=max(1, ttl),
            )

        try:
            await self._take_limit(
                f"auth:otp:request:ip:{self._ip_key(ip_address)}",
                self._settings.otp_ip_requests_per_window,
            )
            await self._take_limit(
                f"auth:otp:request:email:{email_key}",
                self._settings.otp_email_requests_per_window,
            )
        except AuthError:
            await self._delete_if_value(cooldown_key, challenge_id)
            raise

        code = generate_otp_code()
        code_hash = hash_otp(
            challenge_id=challenge_id,
            code=code,
            pepper=self._settings.otp_pepper.get_secret_value(),
        )
        challenge_key = f"auth:otp:challenge:{challenge_id}"
        active_key = f"auth:otp:active:{email_key}"
        previous_challenge = await self._redis.get(active_key)
        expected_active = previous_challenge or "__none__"
        previous_key = (
            f"auth:otp:challenge:{previous_challenge}"
            if previous_challenge
            else "auth:otp:challenge:unused"
        )
        installed = await self._redis.eval(
            _INSTALL_ISSUE_SCRIPT,
            3,
            active_key,
            challenge_key,
            previous_key,
            expected_active,
            challenge_id,
            code_hash,
            email,
            self._settings.otp_ttl_seconds,
        )
        if installed[0] != "installed":
            await self._delete_if_value(cooldown_key, challenge_id)
            raise AuthError(
                code="rate_limited",
                message="A newer code is already active. Wait before trying again.",
                status_code=429,
                retry_after_seconds=self._settings.otp_resend_cooldown_seconds,
            )

        try:
            await self._email_sender.send_otp(
                to=email,
                code=code,
                challenge_id=challenge_id,
                expires_in_seconds=self._settings.otp_ttl_seconds,
            )
        except EmailDeliveryError as error:
            await self._redis.eval(
                _ROLLBACK_ISSUE_SCRIPT,
                3,
                cooldown_key,
                active_key,
                challenge_key,
                challenge_id,
            )
            raise AuthError(
                code="email_delivery_unavailable",
                message="Unable to send a code right now. Try again shortly.",
                status_code=503,
            ) from error

        return IssuedOtp(
            challenge_id=challenge_id,
            expires_in_seconds=self._settings.otp_ttl_seconds,
            resend_after_seconds=self._settings.otp_resend_cooldown_seconds,
        )

    async def reserve(self, *, challenge_id: str, code: str, ip_address: str) -> OtpReservation:
        await self._take_limit(
            f"auth:otp:verify:ip:{self._ip_key(ip_address)}",
            self._settings.otp_verify_attempts_per_ip_window,
        )
        candidate_hash = hash_otp(
            challenge_id=challenge_id,
            code=code,
            pepper=self._settings.otp_pepper.get_secret_value(),
        )
        challenge_key = f"auth:otp:challenge:{challenge_id}"
        email_hint = await self._redis.hget(challenge_key, "email")
        if not email_hint:
            raise invalid_code_error()
        active_key = f"auth:otp:active:{self._email_key(email_hint)}"
        reservation_id = generate_opaque_token()
        result = await self._redis.eval(
            _RESERVE_SCRIPT,
            2,
            challenge_key,
            active_key,
            challenge_id,
            candidate_hash,
            self._settings.otp_max_attempts,
            reservation_id,
        )
        status = result[0]
        if status == "verified":
            return OtpReservation(
                challenge_id=challenge_id,
                email=str(result[1]),
                reservation_id=reservation_id,
            )
        if status == "locked":
            raise AuthError(
                code="code_attempts_exhausted",
                message="Too many incorrect attempts. Request a new code.",
                status_code=429,
                retry_after_seconds=self._settings.otp_resend_cooldown_seconds,
            )
        raise invalid_code_error()

    async def finalize(self, reservation: OtpReservation) -> bool:
        active_key = f"auth:otp:active:{self._email_key(reservation.email)}"
        result = await self._redis.eval(
            _FINALIZE_SCRIPT,
            2,
            f"auth:otp:challenge:{reservation.challenge_id}",
            active_key,
            reservation.reservation_id,
            reservation.challenge_id,
        )
        return bool(int(result))

    async def release(self, reservation: OtpReservation) -> bool:
        active_key = f"auth:otp:active:{self._email_key(reservation.email)}"
        result = await self._redis.eval(
            _RELEASE_SCRIPT,
            2,
            f"auth:otp:challenge:{reservation.challenge_id}",
            active_key,
            reservation.reservation_id,
            reservation.challenge_id,
        )
        return bool(int(result))

    async def verify(self, *, challenge_id: str, code: str, ip_address: str) -> str:
        reservation = await self.reserve(
            challenge_id=challenge_id,
            code=code,
            ip_address=ip_address,
        )
        try:
            if not await self.finalize(reservation):
                raise invalid_code_error()
        except Exception:
            await self.release(reservation)
            raise
        return reservation.email

    async def close(self) -> None:
        await self._redis.aclose()

    async def _delete_if_value(self, key: str, value: str) -> None:
        await self._redis.eval(
            "if redis.call('GET', KEYS[1]) == ARGV[1] then "
            "return redis.call('DEL', KEYS[1]) end return 0",
            1,
            key,
            value,
        )

File: apps/api/src/modules/auth/otp.ts
import type { Redis } from 'ioredis'
import type { Settings } from '../../config.js'
import { EmailDeliveryError, type OtpEmailSender } from '../../email.js'
import { AuthError, invalidCode } from '../../errors.js'
import { generateOpaqueToken, generateOtpCode, hashOtp, keyedIdentifier } from '../../security.js'

export type IssuedOtp = { challengeId: string; expiresInSeconds: number; resendAfterSeconds: number; developmentCode?: string }
export type OtpReservation = { challengeId: string; email: string; reservationId: string }

const fixedWindowScript = `
local current = redis.call('INCR', KEYS[1])
if current == 1 then redis.call('EXPIRE', KEYS[1], ARGV[1]) end
return {current, redis.call('TTL', KEYS[1])}`

const installScript = `
local current = redis.call('GET', KEYS[1])
if ARGV[1] == '__none__' then
  if current then return {'stale'} end
elseif current ~= ARGV[1] then
  return {'stale'}
end
if current then redis.call('DEL', KEYS[3]) end
redis.call('HSET', KEYS[2], 'code_hash', ARGV[3], 'email', ARGV[4], 'attempts', '0', 'status', 'active')
redis.call('EXPIRE', KEYS[2], ARGV[5])
redis.call('SET', KEYS[1], ARGV[2], 'EX', ARGV[5])
return {'installed'}`

// Development issues no email and replaces the current challenge in one operation.
const replaceDevelopmentScript = `
local current = redis.call('GET', KEYS[1])
if current then redis.call('DEL', 'auth:otp:challenge:' .. current) end
redis.call('HSET', KEYS[2], 'code_hash', ARGV[2], 'email', ARGV[3], 'attempts', '0', 'status', 'active')
redis.call('EXPIRE', KEYS[2], ARGV[4])
redis.call('SET', KEYS[1], ARGV[1], 'EX', ARGV[4])
return {'installed'}`

const reserveScript = `
if redis.call('EXISTS', KEYS[1]) == 0 then return {'missing'} end
if redis.call('GET', KEYS[2]) ~= ARGV[1] then return {'missing'} end
if redis.call('HGET', KEYS[1], 'status') == 'reserved' then return {'already_reserved'} end
if redis.call('HGET', KEYS[1], 'code_hash') == ARGV[2] then
  local email = redis.call('HGET', KEYS[1], 'email')
  redis.call('HSET', KEYS[1], 'status', 'reserved', 'reservation', ARGV[4])
  return {'verified', email}
end
local attempts = redis.call('HINCRBY', KEYS[1], 'attempts', 1)
if attempts >= tonumber(ARGV[3]) then
  redis.call('DEL', KEYS[1]); redis.call('DEL', KEYS[2]); return {'locked'}
end
return {'invalid', tostring(attempts)}`

const finalizeScript = `
if redis.call('HGET', KEYS[1], 'reservation') ~= ARGV[1] then return 0 end
redis.call('DEL', KEYS[1])
if redis.call('GET', KEYS[2]) == ARGV[2] then redis.call('DEL', KEYS[2]) end
return 1`

const releaseScript = `
if redis.call('HGET', KEYS[1], 'reservation') ~= ARGV[1] then return 0 end
if redis.call('GET', KEYS[2]) == ARGV[2] then
  redis.call('HSET', KEYS[1], 'status', 'active'); redis.call('HDEL', KEYS[1], 'reservation')
else
  redis.call('DEL', KEYS[1])
end
return 1`

const rollbackIssueScript = `
if redis.call('GET', KEYS[1]) == ARGV[1] then redis.call('DEL', KEYS[1]) end
if redis.call('GET', KEYS[2]) == ARGV[1] then redis.call('DEL', KEYS[2]) end
redis.call('DEL', KEYS[3])
return 1`

const deleteIfValueScript = `
if redis.call('GET', KEYS[1]) == ARGV[1] then return redis.call('DEL', KEYS[1]) end
return 0`

export class OtpService {
  constructor(
    private readonly redis: Redis,
    private readonly emailSender: OtpEmailSender,
    private readonly settings: Settings,
  ) {}

  private emailKey(email: string) { return keyedIdentifier(email, this.settings.rateLimitPepper) }
  private ipKey(ip: string) { return keyedIdentifier(ip, this.settings.rateLimitPepper) }

  private async takeLimit(key: string, maximum: number) {
    const [current, ttl] = await this.redis.eval(
      fixedWindowScript,
      1,
      key,
      this.settings.otpRateLimitWindowSeconds,
    ) as [number, number]
    if (Number(current) <= maximum) return
    throw new AuthError('rate_limited', 'Too many attempts. Wait before trying again.', 429, Math.max(1, Number(ttl)))
  }

  async issue(email: string, ip: string): Promise<IssuedOtp> {
    const emailKey = this.emailKey(email)
    const challengeId = generateOpaqueToken()
    if (this.settings.environment === 'development') {
      const code = generateOtpCode()
      await this.redis.eval(
        replaceDevelopmentScript,
        2,
        `auth:otp:active:${emailKey}`,
        `auth:otp:challenge:${challengeId}`,
        challengeId,
        hashOtp(challengeId, code, this.settings.otpPepper),
        email,
        this.settings.otpTtlSeconds,
      )
      return {
        challengeId,
        expiresInSeconds: this.settings.otpTtlSeconds,
        resendAfterSeconds: 0,
        developmentCode: code,
      }
    }
    const cooldownKey = `auth:otp:cooldown:${emailKey}`
    const acquired = await this.redis.set(cooldownKey, challengeId, 'EX', this.settings.otpResendCooldownSeconds, 'NX')
    if (!acquired) {
      throw new AuthError('rate_limited', 'Wait before requesting another code.', 429, Math.max(1, await this.redis.ttl(cooldownKey)))
    }

    try {
      await this.takeLimit(`auth:otp:request:ip:${this.ipKey(ip)}`, this.settings.otpIpRequestsPerWindow)
      await this.takeLimit(`auth:otp:request:email:${emailKey}`, this.settings.otpEmailRequestsPerWindow)
    } catch (error) {
      await this.redis.eval(deleteIfValueScript, 1, cooldownKey, challengeId)
      throw error
    }

    const code = generateOtpCode()
    const activeKey = `auth:otp:active:${emailKey}`
    const previous = await this.redis.get(activeKey)
    const previousKey = previous ? `auth:otp:challenge:${previous}` : 'auth:otp:challenge:unused'
    const challengeKey = `auth:otp:challenge:${challengeId}`
    let installResult: string[]
    try {
      installResult = await this.redis.eval(
        installScript,
        3,
        activeKey,
        challengeKey,
        previousKey,
        previous ?? '__none__',
        challengeId,
        hashOtp(challengeId, code, this.settings.otpPepper),
        email,
        this.settings.otpTtlSeconds,
      ) as string[]
    } catch (error) {
      await this.redis.eval(deleteIfValueScript, 1, cooldownKey, challengeId)
      throw error
    }
    if (installResult[0] !== 'installed') {
      // The active challenge changed between GET and the atomic install. The
      // request owns this cooldown claim and must release it on a stale issue.
      await this.redis.eval(deleteIfValueScript, 1, cooldownKey, challengeId)
      throw new AuthError('rate_limited', 'A newer code is already active. Wait before trying again.', 429, this.settings.otpResendCooldownSeconds)
    }

    try {
      await this.emailSender.sendOtp({ to: email, code, challengeId, expiresInSeconds: this.settings.otpTtlSeconds })
    } catch (error) {
      await this.redis.eval(rollbackIssueScript, 3, cooldownKey, activeKey, challengeKey, challengeId)
      if (error instanceof EmailDeliveryError) {
        throw new AuthError('email_delivery_unavailable', 'Unable to send a code right now. Try again shortly.', 503)
      }
      throw error
    }
    return { challengeId, expiresInSeconds: this.settings.otpTtlSeconds, resendAfterSeconds: this.settings.otpResendCooldownSeconds }
  }

  async reserve(challengeId: string, code: string, ip: string): Promise<OtpReservation> {
    await this.takeLimit(`auth:otp:verify:ip:${this.ipKey(ip)}`, this.settings.otpVerifyAttemptsPerIpWindow)
    const challengeKey = `auth:otp:challenge:${challengeId}`
    const email = await this.redis.hget(challengeKey, 'email')
    if (!email) throw invalidCode()
    const reservationId = generateOpaqueToken()
    const result = await this.redis.eval(
      reserveScript,
      2,
      challengeKey,
      `auth:otp:active:${this.emailKey(email)}`,
      challengeId,
      hashOtp(challengeId, code, this.settings.otpPepper),
      this.settings.otpMaxAttempts,
      reservationId,
    ) as string[]
    if (result[0] === 'verified') return { challengeId, email: String(result[1]), reservationId }
    if (result[0] === 'locked') throw new AuthError('code_attempts_exhausted', 'Too many incorrect attempts. Request a new code.', 429, this.settings.otpResendCooldownSeconds)
    throw invalidCode()
  }

  async finalize(reservation: OtpReservation) {
    const result = await this.redis.eval(
      finalizeScript,
      2,
      `auth:otp:challenge:${reservation.challengeId}`,
      `auth:otp:active:${this.emailKey(reservation.email)}`,
      reservation.reservationId,
      reservation.challengeId,
    )
    return Number(result) === 1
  }

  async release(reservation: OtpReservation) {
    const result = await this.redis.eval(
      releaseScript,
      2,
      `auth:otp:challenge:${reservation.challengeId}`,
      `auth:otp:active:${this.emailKey(reservation.email)}`,
      reservation.reservationId,
      reservation.challengeId,
    )
    return Number(result) === 1
  }

  async close() { await this.redis.quit() }
}
